scanners/sslyze.py

```diff
@@ -36,15 +37,55 @@
 # Advertise Lambda support
 lambda_support = True
 
+# Advertise fast cache support
+fast_cache_support = True
+
+# The environment dictionary key whose correpsonding value is the fast
+# cache data
+FAST_CACHE_KEY = 'fastcache'
+
+
+def get_mail_server_from_cache(mail_server: str, fastcache: dict) -> Union[dict, None]:
+    """
+    Extract and return data from the fast cache that corresponds to
+    the specified mail server.
+
+    Parameters
+    ----------
+    mail_server : str
+        A string of the form 'hostname:port'.
+
+    fastcache : dict
+        The fast cache dictionary.
+
+    Returns
+    -------
+    dict
+        The data found in the fast cache corresponding to the
+        specified mail server, or None if no such data is found.
+    """
+    hostname_and_port = mail_server.split(':')
+    hostname = hostname_and_port[0]
+    port = hostname_and_port[1]
+    for record in fastcache:
+        if record['starttls_smtp'] and record['hostname'] == hostname and record['port'] == port:
+            return record
```

Comment:
This part is really the thing that sticks out the most: it's iterating through a list and interrogating each value, rather than doing a dict lookup. For something like this, I would expect to see the fastcache itself be a dict with mail hosts as the dict keys. Would that be reasonable? If so, that would be simpler code, and much faster.

Reply:
In order to do that I would have to have the fastcache `dict` available to me in the `scan()` function of `scanners/sslyze.py` and I don't.  `scan()` is called by either `perform_local_scan()` or `perform_lambda_scan()` of the file `scan`, and the fastcache `dict` isn't passed to those functions precisely because we don't want such a (potentially) large `dict` to be pushed up to AWS Lambda for every Lambda execution.

An alternative is to put scanner-specific code in `scan` in lines 333-242 where it updates the fastcache, but that's even uglier.  And it breaks the modularity of the code since `scan` should be agnostic to what is inside the individual `scanners/*.py`.

I get that it would be better and more efficient to structure the fastcache `dict` differently.  I just didn't see a decent way to do that with the way scanners currently operate in `domain-scan`.

One possible solution, which admittedly I haven't thought through completely, might be to add a post-scan lifecycle method to the scanners.  `scan` could pass in the original environment (i.e. without removing the fastcache data structure) and the results just returned by `scan` and let the scanner itself update the fastcache data structure however it sees fit.  Do you think this would address your concerns?  Maybe this is what you had in mind all along and I was just missing the point.

Reply:
Got it, I now completely see the challenge in doing it the way I'm recommending. I guess the logical conclusion here is to either add a lifecycle method just for cache management (which feels heavy) or to let scanners absorb the complexity of knowing what their own data structures were as they iterate through a large cache (which is this PR).

Another option might be to allow each scanner to return a `to_cache` key that is a dict with one or more key/value pairs to toss/merge into the fastcache dict, which the `scan` orchestrator can do without knowing what they mean. This would also dramatically slim the size of the data stored in the fastcache key, saving on memory as well as speed. I imagine it wouldn't slim it to the size where it would be acceptable to pass to each Lambda function - but maybe.

Any thoughts on that? That's the last suggestion I'll throw your way, and if you don't like it, I'm happy to merge this and move on.

Reply:
I think the lifecycle method is the cleanest way to implement it.  I could make it so that the lifecycle method is called if (1) the scanner provides it and (2) the `--in-memory-cache` flag is present when `scan` is called.

I'll implement that now and commit it to this pull request.

Reply:
I ended up creating a CLI flag to _disable_ fast caching, since I think normally users will want it.